src/python/pants/backend/jvm/tasks/reports/junit_html_report.py

```diff
@@ -5,23 +5,93 @@
 from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
                         unicode_literals, with_statement)
 
-import glob
+import collections
+import fnmatch
+import itertools
+import logging
 import os
 import xml.etree.ElementTree as ET
 from abc import abstractmethod
 from functools import total_ordering
 
+from twitter.common.collections import OrderedSet
+
 from pants.base.mustache import MustacheRenderer
 from pants.util import desktop
-from pants.util.dirutil import safe_mkdir_for
+from pants.util.dirutil import safe_mkdir_for, safe_walk
+from pants.util.memo import memoized_property
 from pants.util.meta import AbstractClass
+from pants.util.objects import datatype
 from pants.util.strutil import ensure_binary
 
 
+_LOGGER = logging.getLogger(__name__)
+
+
 @total_ordering
 class ReportTestSuite(object):
   """Data object for a JUnit test suite"""
 
+  class MergeError(Exception):
+    def __init__(self, test_cases):
+      error_message = 'Conflicting test cases are not mergeable:\n\t{}'.format(
+        '\n\t'.join(map(repr, test_cases)))
+      super(ReportTestSuite.MergeError, self).__init__(error_message)
+
+  @classmethod
+  def merged(cls, report_test_suites, error_on_conflict=True, logger=None):
+    """Merges any like-named test suites into one.
+
+    :param report_test_suites: A sequence of test suites to merge results from.
+    :type report_test_suites: :class:`collections.Iterable` of :class:`ReportTestSuite`
+    :param bool error_on_conflict: `True` to raise when two or more test cases in a given test suite
+                                   have the same name but otherwise different attributes.
+    :param logger: An optional logger to use for logging merge conflicts.
+    :type logger: :class:`logging.Logger`
+    :raises: :class:`ReportTestSuite.MergeError` if configured to do so on merge errors.
+    :returns: A single test suite with the result of all the given `report_test_suites` merged.
+    :rtype: iter of :class:`ReportTestSuite`s
+    """
+
+    logger = logger or _LOGGER
+
+    suites_by_name = collections.defaultdict(list)
+    for report_test_suite in report_test_suites:
+      suites_by_name[report_test_suite.name].append(report_test_suite)
+
+    for name, suites in suites_by_name.items():
+      cases_by_name = collections.defaultdict(OrderedSet)
```

Comment:
Question: if there are test-case objects that have the same name and time et al I think using a set here means that they'll be quietly merged into one rather than raising an error. Is that the desired behavior, and could you add a test case for that?

Reply:
Exactly, and the test case is [here](https://github.com/pantsbuild/pants/pull/5257/files#diff-1e8eec41e6a2fa1929bf12eb0f7f419dR108) - note the 1st two xml strings have `suite-a/test-a` with identical properties, testing a no-conflict merge.

Reply:
Hm. So if I had two runs of the same suite, as long as both runs had exactly the same timing, we'd report it as though they'd only run once?

In the example, the first `suite-a` run took `0.01` seconds and the second took `0.02` seconds. When they're merged, the merged suite's timing is `0.02`, not `0.03`. If the test case timings were larger, eg ~30sec per case that could be confusing because the reported runtime and the actual runtime would be noticeably different.

I think that's not a big deal as long as the duplicate test case xml somehow represents the same run, but it might be confusing if that isn't the case.

Reply:
Thanks for pushing on that. If we have two runs of the same suite, our programming assumptions have been proven wrong. And to back that up I've switched to a list.